cpp/src/join/quadtree_point_in_polygon.cu

```diff
@@ -0,0 +1,573 @@
+                               point_y,
+                               poly_offsets,
+                               ring_offsets,
+                               poly_points_x,
+                               poly_points_y,
+                               mr,
+                               stream);
+}
+
+}  // namespace detail
+
+std::unique_ptr<cudf::table> quadtree_point_in_polygon(cudf::table_view const &poly_quad_pairs,
+                                                       cudf::table_view const &quadtree,
+                                                       cudf::column_view const &point_indices,
+                                                       cudf::column_view const &point_x,
+                                                       cudf::column_view const &point_y,
+                                                       cudf::column_view const &poly_offsets,
+                                                       cudf::column_view const &ring_offsets,
+                                                       cudf::column_view const &poly_points_x,
+                                                       cudf::column_view const &poly_points_y,
+                                                       rmm::mr::device_memory_resource *mr)
+{
+  CUSPATIAL_EXPECTS(poly_quad_pairs.num_columns() == 2,
+                    "a quadrant-polygon table must have 2 columns");
+  CUSPATIAL_EXPECTS(quadtree.num_columns() == 5, "a quadtree table must have 5 columns");
+  CUSPATIAL_EXPECTS(point_indices.size() == point_x.size() && point_x.size() == point_y.size(),
+                    "number of points must be the same for both x and y columns");
+  CUSPATIAL_EXPECTS(ring_offsets.size() >= poly_offsets.size(),
+                    "number of rings must be no less than number of polygons");
+  CUSPATIAL_EXPECTS(poly_points_x.size() == poly_points_y.size(),
+                    "numbers of vertices must be the same for both x and y columns");
+  CUSPATIAL_EXPECTS(poly_points_x.size() >= 3 * ring_offsets.size(),
+                    "all rings must have at least 3 vertices");
+  CUSPATIAL_EXPECTS(poly_points_x.type() == poly_points_y.type(),
+                    "polygon columns must have the same data type");
+  CUSPATIAL_EXPECTS(point_x.type() == point_y.type(), "point columns must have the same data type");
+  CUSPATIAL_EXPECTS(point_x.type() == poly_points_x.type(),
+                    "points and polygons must have the same data type");
+
+  if (poly_quad_pairs.num_rows() == 0 || quadtree.num_rows() == 0 || point_indices.size() == 0 ||
+      poly_offsets.size() == 0) {
+    std::vector<std::unique_ptr<cudf::column>> cols{};
+    cols.reserve(2);
+    cols.push_back(cudf::make_empty_column(cudf::data_type{cudf::type_id::UINT32}));
+    cols.push_back(cudf::make_empty_column(cudf::data_type{cudf::type_id::UINT32}));
+    return std::make_unique<cudf::table>(std::move(cols));
+  }
+
+  return detail::quadtree_point_in_polygon(poly_quad_pairs,
```

Comment:
Probably need to sync stream 0 here before returning, rather than deeper down in the functions that take streams.